src/Main.elm

```diff
@@ -1,10 +1,396 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
+import Browser
+import Data.Audience exposing (Audience)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Data.AudienceTree exposing (AudienceTree)
 import Html exposing (Html)
+import Html.Attributes
+import Html.Events
+import Json.Decode exposing (Decoder)
+import Process
+import Task
 
 
-main : Html msg
+type Model
+    = Loading LoadingModel
+    | DecodingError
+    | Loaded LoadedModel
+
+
+type Msg
+    = LoadingMsg LoadingMsg
+    | LoadedMsg LoadedMsg
```

Comment:
Could you please walk me through your reasoning on why you chose this approach (custom type instead of a record; Msgs separated by Model cases they apply to), and perhaps pros and cons that you considered between the alternatives?

Reply:
I use a sum type for the model instead of a product one because here.... well... my model is better represented by a sum type! I cannot display anything to user before I have all the data and the user cannot do anything neither. So "Loading" is a special state.

I thinked about it a few minutes, but I only see benefits with this approach and no drawbacks. Because if all the msgs were "flat", we'd have to do something like:
```
update msg model =
    case msg of
        FolderClicked ->
                 -- it doesn't mean anything to open a non loaded folder

                  if isLoaded model) then
                       ....
                  else 
                      <noOp>
        GotFolders (Ok folders) ->
                  if not (isLoaded model) then
                      ....
                  else
                     <noOp>
       ....
```
leading to a lot of repeating code. But if you ask, you surely see some drawbacks and I'd be happy to learn your point of view about that.

Reply:
It seems OK to me. I was thinking about future request such a "be able reload already loaded audience" but it's relatively easy achievable with you approach 👍